velox/dwio/parquet/reader/ParquetReader.cpp

```diff
@@ -293,7 +293,7 @@ std::shared_ptr<const ParquetTypeWithId> ReaderBase::getParquetColumnInfo(
           std::move(name),
           std::nullopt,
           maxRepeat,
-          maxDefine);
+          maxDefine - 1);
```

Comment:
To fix the incorrect result of `TEST_F(ParquetTableScanTest, array)`.

Reply:
Can you explain a little about why do we need to decrease maxDefine here?  To compensate for the extra elements layer?

Reply:
@Yuhta Thanks for your review!
Without this change, the result of `TEST_F(ParquetTableScanTest, array)` is incorrect as below, in which `3` is missing.

```
Expected 1, got 1
1 extra rows, 1 missing rows
1 of extra rows:
	[1,2]

1 of missing rows:
	[1,2,3]
```

At [ParquetTypeWithId.cpp#L79](https://github.com/facebookincubator/velox/blob/main/velox/dwio/parquet/reader/ParquetTypeWithId.cpp#L79), `maxDefine_ + 1` is sent to arrow LevelInfo ([level_conversion.h#L54-L60](https://github.com/apache/arrow/blob/main/cpp/src/parquet/level_conversion.h#L54-L60)), so I assume the max definition level is not matched with the actual schema pattern, but I'm not quite familiar with that.

Below is the schema pattern of this Parquet. Besides this test, there are several other tests in this PR on different schema patterns, including `optional array with required elements` etc.

![Screenshot from 2023-07-19 00-57-43](https://github.com/facebookincubator/velox/assets/41687378/391665d4-5bec-45d6-b025-5890fe106334)

Reply:
Thanks for the explanation.  I think we need some more time to check all these cases, but this PR is still very valuable and provide these cases.  Let's merge this first and work from there.